File: assignment-01/part_b_imperative.py
# -*- coding: windows-1250 -*-

expenses = [
    {"date": "2024-01-05", "category": "Food",          "amount": 42.50, "description": "Groceries"},
    {"date": "2024-01-07", "category": "Transport",     "amount": 15.00, "description": "Bus pass"},
    {"date": "2024-01-09", "category": "Entertainment", "amount": 60.00, "description": "Concert ticket"},
    {"date": "2024-01-10", "category": "Food",          "amount": 8.75,  "description": "Coffee & snack"},
    {"date": "2024-01-12", "category": "Utilities",     "amount": 120.00,"description": "Electricity bill"},
    {"date": "2024-01-14", "category": "Food",          "amount": 55.20, "description": "Restaurant dinner"},
    {"date": "2024-01-15", "category": "Transport",     "amount": 30.00, "description": "Taxi"},
    {"date": "2024-01-17", "category": "Entertainment", "amount": 14.99, "description": "Streaming subscription"},
    {"date": "2024-01-20", "category": "Food",          "amount": 38.00, "description": "Groceries"},
    {"date": "2024-01-22", "category": "Utilities",     "amount": 45.00, "description": "Internet bill"},
    {"date": "2024-01-25", "category": "Transport",     "amount": 22.00, "description": "Train ticket"},
    {"date": "2024-01-28", "category": "Entertainment", "amount": 25.00, "description": "Book"},
]

#// B1 - total and count
total = 0
count = 0
for e in expenses:
    total += e["amount"]
    count += 1
print(f"Total expenses: {total:.2f}")
print(f"Number of records: {count}")

#// B2 - category breakdown
cat_totals = {}
for e in expenses:
    cat = e["category"]
    amt = e["amount"]
    if cat in cat_totals:
        cat_totals[cat] += amt
    else:
        cat_totals[cat] = amt
print("\nCategory breakdown:")
for cat in sorted(cat_totals.keys()):
    print(f"  {cat:15} : {cat_totals[cat]:.2f}")


# B3 - most and least expensive (no max/min)
max_expense = expenses[0]
min_expense = expenses[0]
for e in expenses:
    if e["amount"] > max_expense["amount"]:
        max_expense = e
    if e["amount"] < min_expense["amount"]:
        min_expense = e
print(f"\nMost expensive : {max_expense['description']} ({max_expense['category']}) — {max_expense['amount']:.2f}")
print(f"Least expensive: {min_expense['description']} ({min_expense['category']}) — {min_expense['amount']:.2f}")
# B4 - expenses above average
sum_avg = 0
cnt_avg = 0
for e in expenses:
    sum_avg += e["amount"]
    cnt_avg += 1
average = sum_avg / cnt_avg

above = []
for e in expenses:
    if e["amount"] > average:
        above.append(f"{e['description']} ({e['amount']:.2f})")

print(f"\nAverage expense: {average:.2f}")
print("Expenses above average:")
for item in above:
    print(f"  - {item}")

File: assignment-01/part_b_imperative.cpp
#include <iostream>
#include <iomanip>
#include <string>
using namespace std;

int main() {

    string dates[] = {
        "2024-01-05", "2024-01-07", "2024-01-09", "2024-01-10",
        "2024-01-12", "2024-01-14", "2024-01-15", "2024-01-17",
        "2024-01-20", "2024-01-22", "2024-01-25", "2024-01-28"
    };

    string categories[] = {
        "Food", "Transport", "Entertainment", "Food",
        "Utilities", "Food", "Transport", "Entertainment",
        "Food", "Utilities", "Transport", "Entertainment"
    };

    double amounts[] = {
        42.50, 15.00, 60.00, 8.75, 120.00, 55.20,
        30.00, 14.99, 38.00, 45.00, 22.00, 25.00
    };

    string descriptions[] = {
        "Groceries", "Bus pass", "Concert ticket", "Coffee & snack",
        "Electricity bill", "Restaurant dinner", "Taxi",
        "Streaming subscription", "Groceries", "Internet bill",
        "Train ticket", "Book"
    };

    int count = 12;

    cout << fixed << setprecision(2);

    double total = 0;
    for (int i = 0; i < count; i++) {
        total = total + amounts[i];
    }

    cout << "Total expenses: " << total << endl;
    cout << "Number of records: " << count << endl;

    string uniqueCategories[] = {"Entertainment", "Food", "Transport", "Utilities"};
    double categoryTotals[] = {0, 0, 0, 0};
    int numCategories = 4;

    for (int i = 0; i < count; i++) {
        for (int j = 0; j < numCategories; j++) {
            if (categories[i] == uniqueCategories[j]) {
                categoryTotals[j] = categoryTotals[j] + amounts[i];
            }
        }
    }

    cout << "\nCategory breakdown:" << endl;
    for (int i = 0; i < numCategories; i++) {
        cout << "  " << uniqueCategories[i] << " : " << categoryTotals[i] << endl;
    }

    int maxIndex = 0;
    int minIndex = 0;

    for (int i = 1; i < count; i++) {
        if (amounts[i] > amounts[maxIndex]) {
            maxIndex = i;
        }
        if (amounts[i] < amounts[minIndex]) {
            minIndex = i;
        }
    }

    cout << "\nMost expensive : " << descriptions[maxIndex]
         << " (" << categories[maxIndex] << ")"
         << " - " << amounts[maxIndex] << endl;

    cout << "Least expensive: " << descriptions[minIndex]
         << " (" << categories[minIndex] << ")"
         << " - " << amounts[minIndex] << endl;

    double sum = 0;
    for (int i = 0; i < count; i++) {
        sum = sum + amounts[i];
    }
    double average = sum / count;

    cout << "\nAverage expense: " << average << endl;

    cout << "Expenses above average:" << endl;
    for (int i = 0; i < count; i++) {
        if (amounts[i] > average) {
            cout << "  - " << descriptions[i]
                 << " (" << amounts[i] << ")" << endl;
        }
    }

    return 0;
}
